### DISCUSSION TITLE
Using PTP with a different ptp daemon or avoiding PTP at all?

### DISCUSSION THREAD

**Discussion by chincheta0815** [mikebrady/shairport-sync#1712] [Q&A]
Hi there,

I am thinking of an implementation of shairport-sync, where I might run into issues with NQPTP as it needs root privileges.
Would it be possible to use a different, system-wide ptp daemon (like ptpd) and use some kind of standard api?
Would it also be possible to even avoid the use of PTP? I understand that sound synching might not be working then, which would be fine... Hot would I tell that shairport-sync?

Thanks in advance for some information on that issue.

**Comment by mikebrady** [comment]
Thanks for the discussion topic. There is a great deal that is unknown about how AirPlay 2 does synchronisation, except that it seems to be based on IEEE 1588 "Precision Clock Synchronization Protocol" ("PTP") and possibly IEEE 802.1AS "Timing and Synchronization for Time-Sensitive Applications". It is not clear how potential clock masters are permitted to be used, for example, and it seems that some unknown custom PTP messages are used to manage power-down clocks.

NQPTP tries to do an end-run on this by monitoring PTP signals coming from the AirPlay source and using them to sync the local clock to the player. It does not normally play an active part in PTP interactions, e.g. it does not normally originate PTP messages, respond to PTP signals, participate in clock mastership elections and similar.

On the other hand, it is lightweight, fairly stable, works on any network connection, and does not need timing hardware.

It seems that the AirPlay source will only send and respond to signals on ports 319 and 320 -- the PTP ports. As you know, they normally require root access. However, I understand that you can instruct Linux (and maybe FreeBSD) to permit access to other users or groups.

NQPTP provides a simple shared-memory interface for a Shairport Sync ("SPS") client, and SPS can send simple commands and status updates to it via UDP port 9000. This was originally modelled on a simplification of the OpenAvnu interface but has very significantly changed.

BTW, classic AirPlay ("AirPlay 1") does not use PTP at all.

**Reply by chincheta0815** [reply]
Wow. That is detailed and as always with Apple quite weird ;o)

So I understand: Airplay 2 needs it no matter if one would create sync groups. NQPTP's port is just a value you chose, so it would probably be possible to set it to another port. I am asking as port 9000 is necessary for Logitech Media Server at least for UPD.

As you already wrote the binding should also be possible to be changed by setcaps in Linux so that a non-root user might start and run NQPTP. As far as I found it might be necessary to have all libraries statically linked into the binary. But that is something to be checked. In the end the setcaps could be set in the systemd service file via `AmbientCapabilities=CAP_NET_BIND_SERVICE` in the `Service` section...
